rust/rustc.bzl

```diff
@@ -25,27 +25,165 @@ def _get_rustc_env(ctx):
+
+    Returns:
+      Returns a struct containing the following fields:
+        transitive_crates:
+        transitive_dylibs:
+        transitive_staticlibs:
+        transitive_libs: All transitive dependencies, not filtered by type.
+        setup_cmd:
+        search_flags:
+        link_flags:
+    """
+    staticlib_filetype = FileType([toolchain.staticlib_ext])
+    dylib_filetype = FileType([toolchain.dylib_ext])
+
+    direct_crates = depset()
+    transitive_crates = depset()
+    transitive_dylibs = depset(order = "topological")  # dylib link flag ordering matters.
+    transitive_staticlibs = depset()
+    for dep in deps:
+        if hasattr(dep, "crate_info"):
+            # This dependency is a rust_library
+            direct_crates += [dep.crate_info]
+            transitive_crates += [dep.crate_info]
+            transitive_crates += dep.depinfo.transitive_crates
+            transitive_dylibs += dep.depinfo.transitive_dylibs
+            transitive_staticlibs += dep.depinfo.transitive_staticlibs
+        elif hasattr(dep, "cc"):
+            # This dependency is a cc_library
+            if not allow_cc_deps:
+                fail("Only rust_library, rust_binary, and rust_test targets can " +
+                     "depend on cc_library")
+
+            transitive_dylibs += dylib_filetype.filter(dep.cc.libs)
+            transitive_staticlibs += staticlib_filetype.filter(dep.cc.libs)
+        else:
+            fail("rust targets can only depend on rust_library or cc_library targets.")
+
+    transitive_libs = depset([c.output for c in transitive_crates]) + transitive_staticlibs + transitive_dylibs
+
+    # Create symlinks pointing to each transitive lib in deps_dir.
+    deps_dir = working_dir + "/" + name + ".deps"
+    setup_cmd = ["rm -rf " + deps_dir + "; mkdir " + deps_dir + "\n"]
+    for lib in transitive_libs:
+        setup_cmd += [_create_setup_cmd(lib, deps_dir, in_runfiles)]
+
+    search_flags = []
+    if transitive_crates:
+        search_flags += ["-L dependency={}".format(deps_dir)]
+    if transitive_dylibs or transitive_staticlibs:
+        search_flags += ["-L native={}".format(deps_dir)]
+
+    link_flags = []
+
+    # nb. Crates are linked via --extern regardless of their crate_type
+    link_flags += ["--extern " + crate.name + "=" + deps_dir + "/" + crate.output.basename for crate in direct_crates]
+    link_flags += ["-l dylib=" + _get_lib_name(lib) for lib in transitive_dylibs.to_list()]
+    link_flags += ["-l static=" + _get_lib_name(lib) for lib in transitive_staticlibs.to_list()]
+
+    return struct(
+        setup_cmd = setup_cmd,
+        search_flags = search_flags,
+        link_flags = link_flags,
+        transitive_crates = transitive_crates,
+        transitive_dylibs = transitive_dylibs,
+        transitive_staticlibs = transitive_staticlibs,
+        transitive_libs = list(transitive_libs),
+    )
+
+_setup_deps = setup_deps
+
+CrateInfo = provider(
```

Comment:
1) Should this be hoisted out on top of the other functions? I don't feel strongly about this but it does feel hidden.
2) Is per-field documentation appropriate? I'm thinking specifically the "type" and "output" fields could use some extra detail

Reply:
1. Sure, it sees enough usage; at some point it felt attached to the rustc_compile_action and I was just shy of splitting out rustc.bzl / rustdoc.bzl

2. Yeah, when this settles down I will add more documentation (though I miss static typing here too).